--- 
title: "SVG Tutorial for spatialHeatmap"
author: "Jianhai Zhang, PhD student at Genetics, Genomics and Bioinformatics, University of California, Riverside (<jzhan067@ucr.edu>; <zhang.jianhai@hotmail.com>) <p/> PI: Dr. Thomas Girke, professor at Department of Botany and Plant Sciences, University of California, Riverside (<thomas.girke@ucr.edu>)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float:
        collapsed: true
        smooth_scroll: true
    toc_depth: 3
    fig_caption: yes
    code_folding: show
    number_sections: true
    self_contained: true
fontsize: 14pt
bibliography: bibtex.bib
package: spatialHeatmap
vignette: |
  %\VignetteIndexEntry{SVG tutorial for spatialHeatmap}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

  Maintainer: Jianhai Zhang  

<style>body { text-align: justify }</style>

```{r setup0, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr); opts_chunk$set(message=FALSE, warning=FALSE)
```

# Summary

The R/Bioconductor package spatialHeatmap is designed for intuitive visualisation of configured data matrix and SVG image provided by users. This tutorial is specifically devised for how to make an SVG image and configure it with the data matrix step by step, which is given with *Aribidopsis thaliana* root tissues as an example. For the usage of this package, refer to the [vignette](https://jianhaizhang.github.io/SVG_tutorial_file/vignette.html). 

To make a custom SVG image, a png image where regions should have clear contours, a data matrix, the SVG editor [Inkscape](https://inkscape.org/) are required. The image editor [GIMP](https://www.gimp.org/) is optional. Inkscape is used to draw the SVG image with the PNG image as a template, and configure the SVG image with the data matrix. The data matrix is used to colour different regions in the SVG image. GIMP could be used to automatically extract polygons for the SVG image depending on the quality of the PNG template. 

To reproduce the results in this tutorial, users can download all the relavant files [here](https://github.com/jianhaizhang/SVG_tutorial_file/tree/master/download) (download an SVG image: hover over the image, right click, and select "Save image as..."; download a PNG image: click the image, click "Download", right click, and select "Save image as..."; download a TXT file: click the file, click "Raw", right click, and select "Save as...".).

# Procedure

```{r, eval=TRUE, echo=FALSE, warnings=FALSE}

  url <- "https://raw.githubusercontent.com/jianhaizhang/SVG_tutorial_file/master/image/"

```
## Format Gene Expression Matrix

As a convention, the gene expression matrix should be normalised and filtered before downstream processing. The expression matrix ([Table 1](#table), @Geng2013-ek) used in this tutorial is already normalised, so normalisation is not covered. The function `filter_data` [@SummarizedExperiment; @data.table; @utils] is designed to filter expression matrix, which builds on `pOverA` and `cv` in the package "genefilter" [@Gentleman2018-xj].   

<a name="name_format"></a>
To plot the spatial heatmaps, the data matrix (replicates should be aggregated in advance) must be formatted in the following way:

**Alternative 1: format the sample/condition names using `filter_data.`**

1. Row and column names of the gene expression matrix (expr) should be gene IDs and sample/conditions respectively.

2. Create a data frame (col.metadata) for the metadata of sample/condition of the gene expression matrix. One column is sample and one column is condition, corresponding to column names of the expression matrix. The description of samples and conditions can only consist of letters, digits, dots, single space, or single underscore. Each sample and condition should be unique.

3. Optional: create a data frame (row.metadata) for the metadata of genes. One column is the gene annotation and one column is the gene, corresponding to row names of the expression matrix.

4. Store the 3 data frames in "SummarizedExperiment" and apply to 'filter_data'. se <- SummarizedExperiment(assays=list(expr=as.matrix(expr)), rowData=row.metadata, colData=col.metadata). See `filter_data` for details.

**Alternative 2: format the sample/condition names without using `filter_data`.**

1. Row and column names of the data matrix should be gene IDs and sample/conditions respectively. 

2. In sample/condition names, a sample name is followed by double underscore "__" then the condition. E.g In "epidermis__140mM_1h" shown in [Table 1](#table), "epidermis" is the sample and "140mM_1h" is the condition. 

3. The column names of sample/condition can only consist of letters, digits, dots, single space, or single underscore. 

4. Each column name must be unique. 

Not all samples in the matrix need to be present in the SVG image, vice versa. Only samples common between the SVG image and data matrix are recognised and coloured. Even a dot, space, underscore, uppercase, or lowercase matters.  

**Format the data matrix using [Alternative 1](#name_format)**  

```{r, filter, eval=TRUE, echo=TRUE, warnings=FALSE }

library(spatialHeatmap); library(data.table); library(SummarizedExperiment)

# Creat the "SummarizedExperiment" class. Refer to the R package "SummarizedExperiment" for more details.
data.path <- system.file("extdata/shinyApp/example", "root_expr_row_gen.txt", package = "spatialHeatmap")   
## The expression matrix, where the row and column names should be gene IDs and sample/conditions respectively. This data matrix is truncated from a GEO dataset (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE46205), which is already normalised.
library(data.table); expr <- fread(data.path, sep='\t', header=TRUE, fill=TRUE)
col.na <- colnames(expr)[-ncol(expr)]; row.na <- as.data.frame(expr[, 1])[, 1]
expr <- as.matrix(as.data.frame(expr, stringsAsFactors=FALSE)[, -1])
rownames(expr) <- row.na; colnames(expr) <- col.na

## Condition metadata is data frame. It has a column of tissues and a column of contidions, which correspond to columns of the data matrix "expr".
col.met.path <- system.file("extdata/shinyApp/example", "col_metadata.txt", package = "spatialHeatmap") 
col.metadata <- read.table(col.met.path, header=TRUE, row.names=NULL, sep='\t', stringsAsFactors=FALSE)

## Row metadata is a data frame. It has a column of row (gene) annotations, which correspond to rows of the data matrix "expr".
row.met.path <- system.file("extdata/shinyApp/example", "row_metadata.txt", package = "spatialHeatmap")
row.metadata <- read.table(row.met.path, header=TRUE, row.names=1, sep='\t', stringsAsFactors=FALSE)

## The expression matrix, row metadata, and column metadata are stored in a "SummarizedExperiment" object. The row metadata is optional while column metadata is mandatory. The column names in the expression matrix are not important, since they are ultimately renewed by column metadata.
library(SummarizedExperiment); expr <- as.matrix(expr); se <- SummarizedExperiment(assays=list(expr=expr), rowData=row.metadata, colData=col.metadata)  
exp <- filter_data(data=se, pOA=c(0, 0), CV=c(0.1, Inf), ann='ann', samples='sample', conditions='condition', dir=NULL) 

# Get the filtered matrix. "filter_data" returns a "SummarizedExperiment" object.
df <- assay(exp)

```

<a name="table"></a>
```{r, table, eval=TRUE, echo=FALSE, warnings=FALSE}

  kable(df[1:3, 1:4], caption='Table 1. Gene expression matrix. Rows and columns are genes and sample/conditions respectively.')

``` 

## Make SVG Images 

### <a name="draw"></a> Make Blank SVG Images by Drawing 

If the contour in the PNG image is not clear, GIMP would generate low-quality SVG images, so in this case one can draw the blank SVG image with Inkscape by using the PNG as a template. Below is an example of drawing only two polygons.

  1. Draw polygons <p/><p/> 
  Open the root PNG image [@Mustroph2009-nu] in Inkscape. The image can be zoomed by press "-" or "+" on the keyboard. Select the "Draw freeahnd lines (F6)" at the left tool bar. Left click once at the first corner of the polygon, move to the second corner and double left click, and so on. Lastly, when drawing the last line click at the first corner to seal the polygon.
  <p/><center><img src="`r paste0(url, "open.png")`" width="100" hspace=20><img src="`r paste0(url, "draw_freehand_line.png")`" hspace=20>)<img src="`r paste0(url, "double_click.png")`"><img src="`r paste0(url, "draw_second_line.png")`"><img src="`r paste0(url, "draw_last_line.png")`"></center><p/>

  2. Align polygons <p/><p/> 
  Select "Edit path by nodes (F2)" from the left tool bar. Draw a large rectangle to select the whole sealed polygon and draw another large rectangle to select all nodes.
  <p/><center><img src="`r paste0(url, "edit_path_tool.png")`"><img src="`r paste0(url, "select_path.png")`"><img src="`r paste0(url, "select_path1.png")`" ><img src="`r paste0(url, "select_node.png")`"><img src="`r paste0(url, "select_node1.png")`"></center><p/>

    Click the "Make selected nodes smooth" in the top tool bar. Drag the edges and handles to make the sealed polygon aligned with the template, then the first polygon is finished. <p/>
  <center><img src="`r paste0(url, "make_selected_nodes_smooth.png")`"><img src="`r paste0(url, "drag_handle.png")`"><img src="`r paste0(url, "first_polygon.png")`"></center><p/>

  3. <a name="shape"></a> Make polygons with shapes <p/><p/>
  Alternatively, the polygons can be made with rectangles. Click "Create rectangles and squares (F4)" at the left tool bar and draw a rectangle in the second template polygon. Select the rectangle and click "Object to Path" under the "Path" tab at the top, then the rectangle becomes an SVG path, which can be edited. Switch cursor to "Select and Transform Objects (F1)" and rotate the rectangle as expected. <p/>

  <p/><center><img src="`r paste0(url, "create_rectangle.png")`" hspace=15><img src="`r paste0(url, "draw_rectangle1.png")`" hspace=15><img src="`r paste0(url, "object_to_path1.png")`"></center><p/>
  <center><img src="`r paste0(url, "select_tool.png")`" hspace=15></center><p/>

  &nbsp; &nbsp; Switch the cursor to "Edit path by nodes (F2)" and select all nodes of the rectangle. Click "Make selected nodes corner" at the top tool bar. Drag the edges and handles to overlay the rectangle path on the template, then the second polygon is finished. <p/>

  <p/><center><img src="`r paste0(url, "select_rectangle_node.png")`" hspace=10><img src="`r paste0(url, "make_node_corner.png")`" hspace=10><img src="`r paste0(url, "drag_rectangle.png")`" hspace=10><img src="`r paste0(url, "second_polygon.png")`"></center><p/>

  In this root image there are many polygons, so it takes too much time to draw them individually. However, GIMP can be used to extract the polygons automatically. <p/>

### Make Blank SVG Images with GIMP

If the png image contains many polygons and their contours are clear (e.g. the root image), GIMP can be used to automatically extract the polygons. <p/>

  4. Open the root png image with GIMP. There are five coloured tissues (skyblue, green, red, yellow, blue). The skyblue is not present in the gene matrix while the green, red, yellow, blue correspond to epidermis, cortex, endodermis, stele in the gene matrix respectively. Enable the "Paths" panel under the "Windows" tab at the top. Right click and select "By Color" under "Select". <p/>
  <p/><center><img src="`r paste0(url, "open1.png")`" hspace=10><img src="`r paste0(url, "path_panel.png")`" hspace=10><img src="`r paste0(url, "by_color.png")`"></center> <p/>

  5. Click on the skyblue tissue, then right click, select "To Path" under "Select", and the paths of all skyblue tissues are extracted in the "Paths" panel. <p/>
     <center><img src="`r paste0(url, "click_first_color.png")`" hspace=10><img src="`r paste0(url, "to_path.png")`" hspace=10><img src="`r paste0(url, "first_path.png")`"></center><p/>

  6. Similarly, exrtact paths of other tissues. In the "Paths" panel set paths visible by enabling the eye symbol, then right click and merge visible paths. Right click and export merged paths, which is a blank SVG image. <p/>
  <center><img src="`r paste0(url, "eye.png")`" hspace="10"><img src="`r paste0(url, "merge.png")`" hspace="10"><img src="`r paste0(url, "export.png")`"></center> <p/>

### Make Blank SVG Images with GIMP and Drawing

The drawing method creates accurate SVG images but it is time-consuming, while the GIMP method is faster but it can generate fused polygons. In this tutorial, the root png image has well-separated polygons and clear contours, so GIMP can produce accurate SVG images. Otherwise the resulting SVG images would have mixed and noise polygons. Considering the pros and cons of two methods, the good practice is to first use GIMP to extract polygons then use drawing method to refine them. Below is an example of an SVG image with fused polygons, which is generated by GIMP.

  7. <a name="layer"></a> Place all SVG paths inside a layer/group <p/><p/>
  Open the blank fused SVG image in Inkscape (the image can be zoomed by press "-" or "+" on the keyboard). Under the "Layers" tab at the top click "Layers", the "Layers" panel will come out on the right, then click "+" at the bottom left corner of the panel to add "Layer 1". 
  <p/><center><img src="`r paste0(url, "layer1.png")`" hspace=10><img src="`r paste0(url, "layer2.png")`" hspace=10><img src="`r paste0(url, "add_layer.png")`"></center><p/>

    Draw a rectangle over the fused SVG graph and cut. Then click on "Layer 1" and paste the fused SVG into "Layer 1" (make sure the "Layer 1" is unlocked by refering to the lock symbol). Open the "XML Editor" from the "Edit" tab at the top, if "\<svg:path id="path 77"\>" is under "\<svg:g id="layer1" inkscape:label="Layer 1"\>", then the fused SVG is inside \"Layer 1\".  

     <p/><center><img src="`r paste0(url, "draw_rectangle.png")`"><img src="`r paste0(url, "click_layer.png")`" hspace="10"><img src="`r paste0(url, "paste_layer.png")`"><img src="`r paste0(url, "to_layer.png")`"><img src="`r paste0(url, "in_layer.png")`"></center> <p/>

  8. Refine and modify the SVG image <p/><p/>
  Draw a ractangle over the image, click "Break Apart" under the "Path" tab, then select the outer noisy rectangle by clicking on its edge. Press "delete" on the keyboard to delete it. <p/>

  <center><img src="`r paste0(url, "break_apart1.png")`" hspace="15"><img src="`r paste0(url, "remove_outer_frame.png")`" hspace="15"></center> <p/> 

  Click the edge of the large fused polygon and move. Use the [drawing method](#draw) to make new polygons (blue) with the fused ones as templates. Delete the fused polygons and move back the new polygons to make the tissue complete. <p/>

  <center><img src="`r paste0(url, "move_path.png")`" hspace=10><img src="`r paste0(url, "aligned_3_polygons.png")`" hspace="10"><img src="`r paste0(url, "new_polygon.png")`"></center><p/>

If a large fused polygon needs to be separated, one can use the eraser tool. Drag the fused polygon away from the tissue, select the eraser tool from the left tool bar, then use the eraser to cut the fused polygon into three independent polygons. Select the cut polygons and click "Break Apart" under the "Path" tab at the top, then the three polygons are separated. Place back the three polygons to make the tissue complete.  

  <center><img src="`r paste0(url, "linked_polygon.png")`" hspace="10"><img src="`r paste0(url, "eraser.png")`" hspace=10><img src="`r paste0(url, "erase1.png")`"><img src="`r paste0(url, "erase3.png")`"></center><p/>

  9. Fill and stroke <p/><p/>
  Under "Object" tab at the top, select "Fill and Stroke...", then the "Fill and Stroke (Shift+Ctrl+F)" panel will come out on the right. Select all polygons by drawing a large rectangle over them. <p/>
  <center><img src="`r paste0(url, "fill_tab.png")`" hspace=10><img src="`r paste0(url, "select_all1.png")`"></center><p/>

    <a name="opacity"></a> Under the "Stroke paint" tab in the fill and stroke panel, select "Flat color". Under the "Stroke style" tab, set the stroke width, e.g.: 1.5 px. <p/>
  <center><img src="`r paste0(url, "stroke_flat_color.png")`" width="250" hspace=10><img src="`r paste0(url, "stroke_width.png")`" width="200"></center><p/>

    Under the "Fill" tab, click "No paint" to get a blank SVG image, which is ready to use in next section. <p/>
  <center><img src="`r paste0(url, "fill_no_paint.png")`" hspace=10></center><p/>

## Configure the SVG Image with Gene Matrix

  In the SVG image, each polygon has a unique ID. To plot spatial heatmaps, these IDs should be exactly replaced with sample names in the gene matrix. Even a dot, space, underscore, uppercase, or lowercase matters. No matter how the blank SVG image is created, it should be [placed inside a layer/group](#layer) before start the following steps.

  10. Group same tissues <p/><p/>
  If multiple polygons belong to the same tissue type, they should be grouped together. The example of grouping epidermis is given below. Open "XML Editor..." under the "Edit" tab at the top, then the "XML Editor (Shift+Ctrl+X)" panel comes out on the right. Click all the epidermis polygons while pressing the "Shift" key, right click, and select "Group". A group should not contain another group.

  <p/><center><img src="`r paste0(url, "select_group.png")`" hspace="30"><img src="`r paste0(url, "group.png")`"></center><p/>

  11. Associate polygons with samples <p/><p/>
  The epidermis group \<svg:g id="987"\> shows up in the XML panel. Click "id", change "g987" to "epidermis", and click "Set", then the new group id is set.  

  <p/><center><img src="`r paste0(url, "click_id.png")`" height="250" width="250" hspace="20"><img src="`r paste0(url, "set_id.png")`" height="250" width="250"></center><p/> 

In the "Fill and Stroke (Shift+Ctrl+F)" panel, select "Flat color" under the "Fill" tag, then specify a color for epidermis.      
<center><img src="`r paste0(url, "flat_color.png")`"></center><p/>  

Group other tissues, set ids and colours.   
    <center><img src="`r paste0(url, "colored.png")`"></center><p/>

  12. Polygon orders <p/><p/>
  The polygons are stacked over each other according to their orders in the "XML Editor", so the first polygon might be invisible because it could be covered by the second, third, and so on. For instance, in the brain SVG image [@brain; @hipp], the grey outline polygon (the path "rect5480") is the first one, and partially covered by other polygons. Therefore, users should drag and organise the paths in expected order. <p/>
  <center><img src="`r paste0(url, "brain1.png")`" hspace=15><img src="`r paste0(url, "brain2.png")`"></center><p/>

## Add Text to Label Tissues

  Users can add text to label tissues. Basically, the text is first typed in with the text tool and then the text object is coverted to paths. Next paths are added into the polygon group of target tissue and filled with the same tissue colour. Below is the example of adding text to the epidermis tissue.

  13. Creat text paths <p/><p/>
  Select "Creat and select text objects (F8)" from the left tool bar, drag a text box, and type epidermis. Click on the text object and convert it to path. 
  <center><img src="`r paste0(url, "creat_text.png")`" hspace=10><img src="`r paste0(url, "text_box.png")`" hspace=10><img src="`r paste0(url, "type.png")`" hspace=10><img src="`r paste0(url, "object_to_path1.png")`"></center><p/>

  14. Text fill and stroke <p/><p/>
  Click on the text paths and fill them with the same colour of epidermis using "Pick colors from image (F7)" from the left tool bar. In the "Fill and Stroke" panel, set the stroke style.  
  <center><img src="`r paste0(url, "pick_color_tool.png")`" hspace=10><img src="`r paste0(url, "epidermis_green.png")`" hspace=20></center><p/>

  15. Add text paths to tissue group <p/><p/>
  Click and cut the text, then double click epidermis (green) to enter the group.  
  <center><img src="`r paste0(url, "enter_group.png")`"></center><p/>

    Paste the text anywhere then the text is inside the epidermis group as a "text group". Move/resize the text group as expected. Right click and ungroup the text.  
  <center><img src="`r paste0(url, "ungroup_text.png")`" hspace=5></center><p/>

    All the letters are in the epidermis group as individual paths, which can be seen in the XML Editor.  
  <center><img src="`r paste0(url, "ungroup_text1.png")`"></center><p/>

  To add a pointer, [draw a rectangle](#shape) and convert it to path, fill it with the same style as epidermis. Move, rotate and resize it as expected.  

  <center><img src="`r paste0(url, "drag_pointer.png")`" height="150" width="150"></center><p/>

  Clicking any of other tissues will group the text, pointer, and epidermis together, which can be confirmed by dragging epidermis to see they are moving as a whole.  

  <center><img src="`r paste0(url, "drag_new_group.png")`"></center><p/>

  Similarly, add text to label other tissues. 
  <center><img src="`r paste0(url, "root_R.png")`"></center><p/>

  17. Save the final SVG image, which is ready to use in the spatialHeatmap. The name of saved SVG image can only consists of letters, digits, underscores. E.g. "root_cross_final.svg" is acceptable while "root_cross_final(copy).svg" will give errors. <p/>

## Plot the spatial heatmaps

 In Figure 1, user defined tissue regions are coloured by the expression profile of a target gene under different contidions. For example, in [Table 1](#table) the expression value of gene "PSAC" in "epidermis" under the condition "140mM_1h" is "2.457910", after mapping the "epidermis" is coloured "dark yellow", corresponsing to "2.457910" in the colour scale.

```{r, shm, eval=TRUE, echo=TRUE, warnings=FALSE, fig.align="center", fig.cap=('Figure 1. Spatial Heatmaps of the target gene \"PSAC\". User defined tissue regions are coloured by the expression profile of the target gene.')}

  # The path of the formatted SVG image as shown above.  
  svg.path <- system.file("extdata/shinyApp/example", "root_cross_final.svg", package = "spatialHeatmap")
  # Mapping of the target gene "PSAC" to the SVG image. Multiple target genes can be mapped at the same time. Here only one is mapped. "exp" is the filtered expression matrix as shown above.
  spatial_hm(svg=svg.path, data=exp, samples=NULL, conditions=NULL, pOA=c(0, 0), CV=c(0, 100), ID="PSAC", col.com=c("yellow", "blue", "purple"), width=1, height=1, sub.title.size=9.5, layout="gene", ncol=3)

```


## Troubleshooting

  1. Make sure the coloumn names are formatted in the right way.
  2. Make sure the target tissue names are identical between data matrix and SVG. Even a dot, space, underscore, uppercase, or lowercase matters.
  3. Make sure all the paths, groups are placed in a container group as a whole in the XML editor.
  4. Except for the container group/layer, make sure other groups contain only paths, no groups.
  5. If some grouped polygons are absent in the spatial heatmaps, try to ungroup and regroup them in Inkscape.

# Reference


